### PR TITLE
Implement OAuth 2.0 Server Metadata (RFC 8414)

### LINKED ISSUE #54: Epic: OAuth 2.0 Authorization Server Metadata

### PR DESCRIPTION
**Pull request by Kehrlann**
See https://tools.ietf.org/html/rfc8414

Closes #54

### CODE REVIEW

**Comment by Kehrlann** [comment]
This is still a draft. Remaining todos are:
- Integration tests
- General polishing
- Make sure test coverage is appropriate
- Ensuring consistency with ab591d (polish commit for OIDC Provider Configuration)

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AbstractAuthorizationServerMetadata.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+	/**
+	 * An abstract builder for subclasses of {@link AbstractAuthorizationServerMetadata}.
+	 */
+	public static abstract class AbstractBuilder<T extends AbstractAuthorizationServerMetadata, B extends AbstractBuilder<T, B>> {
+
+		protected final Map<String, Object> claims = new LinkedHashMap<>();
+
+		protected AbstractBuilder() { }
+
+		protected abstract B getThis();	// avoid unchecked casts in builder when doing "return (T) this;"
```
Comment (Kehrlann) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AbstractAuthorizationServerMetadata.java:107]: See https://stackoverflow.com/questions/17164375/subclassing-a-java-builder-class/34741836#34741836

File: oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerMetadataTests.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+/**
+ * Integration tests for OAuth 2.0 Authorization Server Metadata.
+ *
+ * @author Daniel Garnier-Moiroux
+ */
+public class OAuth2AuthorizationServerMetadataTests {
```
Comment (Kehrlann) [inline oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerMetadataTests.java:44]: I am not convinced by the name. These are the integration tests; there's a name conflict with those of the domain object, which are in `org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationServerMetadataTests` .

File: oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerMetadataTests.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+	@EnableWebSecurity
+	@Import(OAuth2AuthorizationServerConfiguration.class)
+	static class AuthorizationServerConfigurationWithIssuer extends AuthorizationServerConfiguration {
+
+		@Bean
+		ProviderSettings providerSettings() {
+			return new ProviderSettings().issuer(issuerUrl);
+		}
+	}
```
Comment (Kehrlann) [inline oauth2-authorization-server/src/test/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/authorization/OAuth2AuthorizationServerMetadataTests.java:96]: I didn't replicate the "loadContext" tests found in `OidcTests` as they would basically exercise the exact same failure paths (e.g. configuration error with an issuer that's an invalid URL)

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AbstractAuthorizationServerMetadata.java [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+					.filter(entry -> entry.getValue() != null)
+					.filter(entry -> Collection.class.isAssignableFrom(entry.getValue().getClass()))
+					.filter(entry -> ((Collection<?>) entry.getValue()).size() == 0)
+					.map(Map.Entry::getKey)
+					.collect(Collectors.toSet());
+
+			for (String claimToRemove : claimsToRemove) {
+				this.claims.remove(claimToRemove);
+			}
+		}
```
Comment (Kehrlann) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AbstractAuthorizationServerMetadata.java:307]: This mutates the `claims` field ; I dislike it a bit. Subclasses will implement their `build` methods like this:

```java
public XXX build() {
	validateCommonClaims();
	validateXXXClaims();
	removeEmptyClaims();
	return new XXX(this.claims);
}

```

The alternative I considered was a more functional `protected static Map<String, Object> omitEmptyClaims(Map<String, Object> claims)`, that subclasses could use like so:

```java
public XXX build() {
	validateCommonClaims();
	validateXXXClaims();
	return new XXX(omitEmptyClaims(this.claims));
}
```

However the `validateClaims` throw runtime exceptions, so I stuck `removeEmptyClaims`.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AbstractAuthorizationServerMetadata.java [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+			claimsConsumer.accept(this.claims);
+			return getThis();
+		}
+
+		/**
+		 * Creates the {@link AbstractAuthorizationServerMetadata}.
+		 *
+		 * @return the {@link AbstractAuthorizationServerMetadata}
+		 */
+		public abstract T build();
```
Comment (Kehrlann) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AbstractAuthorizationServerMetadata.java:272]: I prefer leaving this one abstract, the implementation is small enough that it feels ok to have it in the subclasses.

For reference both subclasses follow the same implementation:
- validate common claims
- validate subclass-specific claims (oidc or oauth 2.0 server metadata)
- remove empty lists from claims
- build

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationServerMetadata.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+ * The claims are defined by the OAuth 2.0 Authorization Server Metadata
+ * specification (RFC 8414).
+ *
+ * @author Daniel Garnier-Moiroux
+ * @since 0.1.0
+ * @see AbstractAuthorizationServerMetadata
+ * @see OAuth2AuthorizationServerMetadataClaimAccessor
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc8414#section-3.2">3.2. Authorization Server Metadata Response</a>
+ */
+public final class OAuth2AuthorizationServerMetadata extends AbstractAuthorizationServerMetadata
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationServerMetadata.java:42]: Rename to `OAuth2AuthorizationServerConfiguration`

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AbstractAuthorizationServerMetadata.java [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+ * either in OpenID Connect Discovery 1.0 or OAuth 2.0 Authorization Server Metadata.
+ * It contains a set of claims about the Provider's configuration.
+ *
+ * @author Daniel Garnier-Moiroux
+ * @since 0.1.0
+ * @see AuthorizationServerMetadataClaimAccessor
+ * @see <a target="_blank" href="https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationResponse">4.2. OpenID Provider Configuration Response</a>
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc8414#section-3.2">3.2. Authorization Server Metadata Response</a>
+ */
+public abstract class AbstractAuthorizationServerMetadata implements AuthorizationServerMetadataClaimAccessor, Serializable {
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AbstractAuthorizationServerMetadata.java:49]: Rename to `AbstractAuthorizationServerConfiguration`

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/http/converter/OAuth2AuthorizationServerMetadataHttpMessageConverter.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+/**
+ * A {@link HttpMessageConverter} for an {@link OAuth2AuthorizationServerMetadata OAuth 2.0 Authorization Server Metadata Response}.
+ *
+ * @author Daniel Garnier-Moiroux
+ * @since 0.1.0
+ * @see AbstractHttpMessageConverter
+ * @see OAuth2AuthorizationServerMetadata
+ */
+public class OAuth2AuthorizationServerMetadataHttpMessageConverter
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/http/converter/OAuth2AuthorizationServerMetadataHttpMessageConverter.java:49]: Rename to `OAuth2AuthorizationServerConfigurationHttpMessageConverter`

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationServerMetadataEndpointFilter.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * A {@code Filter} that processes OAuth 2.0 Authorization Server Metadata Requests.
+ *
+ * @author Daniel Garnier-Moiroux
+ * @since 0.1.0
+ * @see ProviderSettings
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc8414#section-3">3. Obtaining Authorization Server Metadata</a>
+ */
+public class OAuth2AuthorizationServerMetadataEndpointFilter extends OncePerRequestFilter {
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationServerMetadataEndpointFilter.java:44]: Rename to `OAuth2AuthorizationServerConfigurationEndpointFilter`

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AuthorizationServerMetadataClaimAccessor.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+ * its configuration, used either in OpenID Connect Discovery 1.0 or OAuth 2.0 Authorization
+ * Server Metadata.
+ *
+ * @author Daniel Garnier-Moiroux
+ * @since 0.1.0
+ * @see ClaimAccessor
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc8414#section-2">2. Authorization Server Metadata</a>
+ * @see <a target="_blank" href="https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata">3. OpenID Provider Metadata</a>
+ */
+public interface AuthorizationServerMetadataClaimAccessor extends ClaimAccessor {
```
Comment (jgrandja) [inline oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AuthorizationServerMetadataClaimAccessor.java:33]: I feel like we should merge this into `OAuth2AuthorizationServerMetadataClaimAccessor`. Then we can remove `AuthorizationServerMetadataClaimAccessor`, `AuthorizationServerMetadataClaimNames` and `AbstractAuthorizationServerMetadata`. This change will simplify the hierarchy.

Although `revocation_endpoint`, `revocation_endpoint_auth_methods_supported` and `code_challenge_methods_supported` are not defined in OIDC Provider Metadata, I believe an OIDC provider will support these configurations either way.

What are your thoughts on this proposed change?

**Review by jgrandja — CHANGES_REQUESTED** [review]
Thanks for the PR @Kehrlann ! Please see review comments.

In addition to the comments, please ensure copyright year is `2020-2021` and `@since` is `0.1.1` for all files.

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AuthorizationServerMetadataClaimAccessor.java (hunk 10 above)
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AuthorizationServerMetadataClaimAccessor.java:33]: 👍  definitely agree with the proposed change

Quick research:
- Okta supports those three claims in their OIDC configuration
- [Google](https://accounts.google.com/.well-known/openid-configuration) supports `revocation_endpoint` and `code_challenge_methods_supported`
- [Microsoft](https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration) supports none of them

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationServerMetadataEndpointFilter.java (hunk 9 above)
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationServerMetadataEndpointFilter.java:44]: Should we also rename
`DEFAULT_OAUTH2_AUTHORIZATION_SERVER_METADATA_ENDPOINT_URI` 
to
`DEFAULT_OAUTH2_AUTHORIZATION_SERVER_CONFIGURATION_ENDPOINT_URI`

in that class ?

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AuthorizationServerMetadataClaimAccessor.java (hunk 10 above)
Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AuthorizationServerMetadataClaimAccessor.java:33]: One caveat, removing `AbstractAuthorizationServerMetadata`, using `OAuth2AuthorizationServerConfiguration` instead in `OidcProviderConfiguration extends OAuth2AuthorizationServerConfiguration` 

Is only possible if we make `OAuth2AuthorizationServerConfiguration` not final. And I believe we want our model classes to be final?

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationServerMetadataEndpointFilter.java (hunk 9 above)
Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationServerMetadataEndpointFilter.java]: Yes please

File: oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AuthorizationServerMetadataClaimAccessor.java (hunk 10 above)
Comment (jgrandja) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AuthorizationServerMetadataClaimAccessor.java:33]: Our typical design process is to make things final (or restrict access) at the start. However, if there is need to open things up than we do only if it makes sense. This is one of those instances so it's ok to make `OAuth2AuthorizationServerConfiguration` non-final BUT make the constructor `protected`.

Comment (Kehrlann) [reply oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/core/AuthorizationServerMetadataClaimAccessor.java:33]: So I've tried this, unfortunately the type hierarchy with generic builders does not allow to have this kind of hierarchy:

```
- OAuth2AuthorizationServerConfiguration
  - Builder (inner class)
  - static builder() method, returning a Builder
  - OidcProviderConfiguration extends OAuth2AuthorizationServerConfiguration
    - OidcBuilder (inner class, extends Builder) 
    - static builder() method, returning an OidcBuilder
```

Problems are:
- If `OidcBuilder` extends the concrete class `OAuth2AuthorizationServerConfiguration.Builder`, the builder pattern won't work as, for example, `builder.issuer("some string")` would always return the parent `Builder` type. I don't think we can get away without a base abstract builder, see [this Stackoverflow thread for reference](https://stackoverflow.com/questions/21086417/builder-pattern-and-inheritance) for reference.
- If both Builder and OidcBuilder inherit a common, generic, AbstractBuilder, then OidcProviderConfiguration.builder() return type becomes incompatible with that of `OAuth2AuthorizationServerConfiguration.builder()`.

If we really want to get rid of the `AbstractAuthorizationServerConfiguration`, we could have the abstract builder and default claims in the `AuthorizationServerMetadataClaimAccessor` interface.

**Comment by jgrandja** [comment]
Thanks for all the work @Kehrlann ! This is now merged.

I decided to rename `OAuth2AuthorizationServerConfiguration` back to the original `OAuth2AuthorizationServerMetadata`. It made more sense.

I also made a few other changes in a follow-up polish commit. Please let me know if you have any questions in regards to the changes. Thanks!

**Pull request merged**